sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala

```diff
@@ -91,10 +91,28 @@ class InMemoryFileIndex(
   }
 
   private def refresh0(): Unit = {
-    val files = listLeafFiles(rootPaths)
-    cachedLeafFiles =
-      new mutable.LinkedHashMap[Path, FileStatus]() ++= files.map(f => f.getPath -> f)
-    cachedLeafDirToChildrenFiles = files.toArray.groupBy(_.getPath.getParent)
+    if (sparkSession.sessionState.conf.scanSubDirEnabled) {
+      val (files, discovered) = listLeafFilesScanSubDirEnable(rootPaths)
```

Comment:
private def refresh0(): Unit = {
    val files = listLeafFiles(rootPaths)
    cachedLeafFiles =
      new mutable.LinkedHashMap[Path, FileStatus]() ++= files.map(f => f.getPath -> f)
    cachedLeafDirToChildrenFiles = if (sparkSession.sessionState.conf.recursiveFileLookup) {
      recursiveDirChildrenFiles(files)
    } else {
      files.toArray.groupBy(_.getPath.getParent)
    }
    cachedPartitionSpec = null
  }

  private def recursiveDirChildrenFiles(files: mutable.LinkedHashSet[FileStatus])
  : Map[Path, Array[FileStatus]] = {
    val rootPathSet = rootPaths.toSet
    val reorganized = mutable.LinkedHashMap[Path, mutable.LinkedHashSet[FileStatus]]()
    files.foreach { f => // f is a file, not a directory.
      var parent = f.getPath.getParent
      while(parent != null && !rootPathSet.contains(parent)) {
        reorganized.getOrElseUpdate(parent, mutable.LinkedHashSet[FileStatus]()).add(f)
        parent = parent.getParent
      }
    }
    reorganized.mapValues(_.toArray).toMap
  }

Reply:
root path:
/user/hive/ssb.db/lineorders/


/user/hive/ssb.db/lineorders/part1=202210/a1.parquet
/user/hive/ssb.db/lineorders/part1=202210/a2.parquet
/user/hive/ssb.db/lineorders/part1=202210/tez1/b1.parquet
/user/hive/ssb.db/lineorders/part1=202210/tez1/b2.parquet
/user/hive/ssb.db/lineorders/part1=202210/z1.parquet

before change:

/user/hive/ssb.db/lineorders/part1=202210/ -> a1.parquet,a2.parquet,z1.parquet
/user/hive/ssb.db/lineorders/part1=202210/tez1/ -> b1.parquet,b2.parquet

after change:

/user/hive/ssb.db/lineorders/part1=202210/ -> a1.parquet,a2.parquet,b1.parquet,b2.parquet,z1.parquet
/user/hive/ssb.db/lineorders/part1=202210/tez1/ -> b1.parquet,b2.parquet

-----


/user/hive/ssb.db/lineorders/part1=202211/tez1/b1.parquet
/user/hive/ssb.db/lineorders/part1=202211/tez1/b2.parquet

before change:
/user/hive/ssb.db/lineorders/part1=202211/tez1/ -> b1.parquet,b2.parquet

after change:
/user/hive/ssb.db/lineorders/part1=202211/ -> b1.parquet,b2.parquet
/user/hive/ssb.db/lineorders/part1=202211/tez1/ -> b1.parquet,b2.parquet

----